src/main/java/org/rumbledb/compiler/VariableDependenciesVisitor.java

```diff
@@ -0,0 +1,524 @@
+import org.rumbledb.expressions.flowr.ReturnClause;
+import org.rumbledb.expressions.flowr.SimpleMapExpression;
+import org.rumbledb.expressions.flowr.WhereClause;
+import org.rumbledb.expressions.module.FunctionDeclaration;
+import org.rumbledb.expressions.module.Prolog;
+import org.rumbledb.expressions.module.VariableDeclaration;
+import org.rumbledb.expressions.postfix.DynamicFunctionCallExpression;
+import org.rumbledb.expressions.postfix.PredicateExpression;
+import org.rumbledb.expressions.primary.ContextItemExpression;
+import org.rumbledb.expressions.primary.FunctionCallExpression;
+import org.rumbledb.expressions.primary.InlineFunctionExpression;
+import org.rumbledb.expressions.primary.NamedFunctionReferenceExpression;
+import org.rumbledb.expressions.primary.VariableReferenceExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpressionVar;
+
+
+/**
+ * This visitor resolves dependencies between variable and function declarations.
+ * 
+ * If a variable $x depends on a variable $y, then $y must be evaluated before $x.
+ * 
+ * Example:
+ * 
+ * declare variable $y := 1;
+ * declare variable $x := $y;
+ * 
+ * If a variable $x depends on a function f, then f's closure must be built before $x is evaluated.
+ * 
+ * declare function f() { 1 };
+ * declare variable $x := f();
+ * 
+ * If a function f depends on a variable $x, then $x must be evaluated before f's closure is built.
+ *
+ * declare variable $x := 1;
+ * declare function f() { $x };
+ * 
+ * Note that a function cannot depend on a function, as mutually recursive calls are allowed.
+ * 
+ * Once all dependencies have been determined, the visitor builds a DAG, builds a topological ordering
+ * thereof, and re-sorts declarations in the prolog for further processing by other visitors.
+ * 
+ */
+public class VariableDependenciesVisitor extends AbstractNodeVisitor<Void> {
+
+    private RumbleRuntimeConfiguration rumbleRuntimeConfiguration;
+
+    /**
+     * Input variable dependencies are lists of variables and functions that an expression depends on.
+     */
+    Map<Node, Set<String>> inputVariableDependencies;
+    /**
+     * Output variable dependencies are lists of variables in the tuples that a clause produces.
+     */
+    Map<Node, Set<String>> outputVariableDependenciesForClauses;
+
+    /**
+     * Builds a new visitor.
+     * 
+     * @param rumbleRuntimeConfiguration the configuration. This is used for trigerring or not debug output.
+     */
+    VariableDependenciesVisitor(RumbleRuntimeConfiguration rumbleRuntimeConfiguration) {
+        this.outputVariableDependenciesForClauses = new HashMap<>();
+        this.inputVariableDependencies = new HashMap<>();
+        this.rumbleRuntimeConfiguration = rumbleRuntimeConfiguration;
+    }
+
+    private void addInputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null || variables.isEmpty()) {
```

Comment:
The code looks great as it is. Thanks a lot for addressing the mentioned issues. 

I have one last suggestion to really push the implementation safety to its limits. We have ensured that getter methods always return emptySet for null cases. Most of the add method calls receive values from the said getter methods. If I'm not missing a case, the remaining add methods are called with variable names extracted from expressions for which we know for sure that the variable name exists and is non-null. 

As such, I feel that we can permit the usage of null parameters to the add methods as in the current ecosystem, such an add call should not happen. In the future, this could hide certain bugs by simply being too lenient for the null parameter case. How would you feel about throwing an 'OurBadException' if null parameter is observed in the add methods.